### PR TITLE
fix: lifi l1 gas estimation for optimism

### PR DESCRIPTION
**Pull request by woodenfurniture**
## Description

Routes lifi quotes though our gas estimation so optimism l1 gas estimates are taken into account. This allows max trades to be more reliable because the gas estimate isnt missing the l1 gas fee component which is roughly $0.35.

Please note
- max trades rely on gas estimates to determine the maximum sell value and this is inherently unreliable
- our backend currently sends an error when estimating gas for expensive lifi optimism trades, which means quotes will be blocked a lot of the time --- this is better than whats in prod because it prevent a failed trade.

## Pull Request Type

- [x] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [ ] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

closes #

## Screenshots (if applicable)
Actual fee was $0.97, we're within a few cents of the actual total network fee which includes l1 gas
https://optimistic.etherscan.io/tx/0xc762b0031e4f42f19c13e2ddb157a6930d95a28787ab90ff53169af61ff5cfcd
```
Amount paid to the sequencer + aggregated l1 fees to process the transaction.
```
image

We estimated 0.000505 ETH:
image

The completed transaction in etherscan (0.000524 ETH):
image

Metamask (0.000524 ETH):
image

### CODE REVIEW

File: src/lib/swapper/swappers/LifiSwapper/utils/getIntermediaryTransactionOutputs/getIntermediaryTransactionOutputs.ts [hunk 1]
```diff
@@ -10,11 +10,9 @@ import type { LifiStepSubset, StepSubset } from './types'
 const isLifiStep = (step: StepSubset): step is LifiStepSubset => step.type === 'lifi'
 
 export const getIntermediaryTransactionOutputs = (
-  selectedLifiRouteSteps: StepSubset[],
+  step: StepSubset,
 ): AmountDisplayMeta[] | undefined => {
-  const tradeSteps = selectedLifiRouteSteps.flatMap(step =>
-    isLifiStep(step) ? step.includedSteps : step,
-  )
+  const tradeSteps = isLifiStep(step) ? step.includedSteps : [step]
```
Comment (0xApotheosis) [inline src/lib/swapper/swappers/LifiSwapper/utils/getIntermediaryTransactionOutputs/getIntermediaryTransactionOutputs.ts:15]: As we are in the LiFi swapper domain here, when would this ever not be a LifI step?

File: src/lib/swapper/swappers/LifiSwapper/utils/transformLifiFeeData/transformLifiFeeData.ts [hunk 2]
```diff
@@ -1,77 +1,57 @@
...
-            symbol: asset?.symbol ?? token.symbol,
-          },
-          requiresBalance: true,
-        }
-      } else {
-        acc[assetId].amountCryptoBaseUnit = bn(acc[assetId].amountCryptoBaseUnit)
-          .plus(amount)
-          .toString()
+  const protocolFees = allPayableFeeCosts.reduce<Record<AssetId, ProtocolFee>>((acc, feeCost) => {
+    const { amount, token } = feeCost
```
Comment (0xApotheosis) [inline src/lib/swapper/swappers/LifiSwapper/utils/transformLifiFeeData/transformLifiFeeData.ts:35]: We might want to destructure this as below for clarity:

```suggestion
    const { amount: amountCryptoBaseUnit, token } = feeCost
```

File: src/lib/swapper/swappers/LifiSwapper/utils/getNetworkFeeCryptoBaseUnit/getNetworkFeeCryptoBaseUnit.ts [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    const optimismGasOracleAbi: AbiItem[] = [
+      {
+        inputs: [{ internalType: 'bytes', name: '_data', type: 'bytes' }],
+        name: 'getL1Fee',
+        outputs: [{ internalType: 'uint256', name: '', type: 'uint256' }],
+        stateMutability: 'view',
+        type: 'function',
+      },
+    ]
+    const web3 = new Web3(provider)
```
Comment (0xApotheosis) [inline src/lib/swapper/swappers/LifiSwapper/utils/getNetworkFeeCryptoBaseUnit/getNetworkFeeCryptoBaseUnit.ts:58]: We might want to use `getWeb3ProviderByChainId(chainId)` here to access the existing singleton web3 instance.

**Review by 0xApotheosis — APPROVED** [review]
Implementation is sane, a few non-blocking comments.

I wasn't able to functionally test due to an issue with my local wallet state rugging me rn.

**Review by 0xApotheosis — APPROVED** [review]

**Review by gomesalexandre — COMMENTED** [review]
Functional pass - tested by mocking `isBelowMinSellAmount` to false and `MIN_AMOUNT_THRESHOLD_USD_HUMAN` to 1 to be able to test smallish Li.Fi trades

While testing this, also noticed we were missing the conversion to Gnosis native AssetId in `lifiTokenToAssetId` - https://github.com/shapeshift/web/pull/4616:

```diff
diff --git a/src/lib/swapper/swappers/LifiSwapper/utils/lifiTokenToAssetId/lifiTokenToAssetId.ts b/src/lib/swapper/swappers/LifiSwapper/utils/lifiTokenToAssetId/lifiTokenToAssetId.ts
index d5d91923e..4d187f930 100644
--- a/src/lib/swapper/swappers/LifiSwapper/utils/lifiTokenToAssetId/lifiTokenToAssetId.ts
+++ b/src/lib/swapper/swappers/LifiSwapper/utils/lifiTokenToAssetId/lifiTokenToAssetId.ts
@@ -49,6 +49,11 @@ export const lifiTokenToAssetId = (lifiToken: Token): AssetId => {
           assetReference: ASSET_REFERENCE.Polygon,
           assetNamespace: ASSET_NAMESPACE.slip44,
         }
+      case CHAIN_REFERENCE.GnosisMainnet:
+        return {
+          assetReference: ASSET_REFERENCE.Gnosis,
+          assetNamespace: ASSET_NAMESPACE.slip44,
+        }
       default:
         throw new SwapError(`[lifiTokenToAssetId] chainId '${lifiToken.chainId}' not supported`, {
           code: SwapErrorType.UNSUPPORTED_CHAIN,
```

### Paranoia "regression" (ish, since XDAI native asset mapping isn't currently present) test of a swap to OP via hop - no L1 fees since the fees are taken in XDAI on the sell side only

image
image
image

Note, despite receiving the ETH on Optimism, and the polling to `/v1/status` properly returning us complete Tx, the trade didn't update to completed, and the Tx history didn't update, though that seems to be a Gnosis ws stability issue

image
image

This is with prod env, and was happy after a refresh

image

### Max ETH on Optimism -> AVAX 

#### Gas and protocol fees are taken into account for the max

image
image

#### Was greeted by a JSON-RPC error and a Tx locally failing in MM

image

Note this may or may not be related to the monkey patching above

### Non-max ETH on Optimism -> AVAX

image
image

### Min expected is correctly showing the assets from the intermediary steps which may fail

image
image

Another thing I've noticed and we may want to report to Li.Fi is despite the Tx being completed on Layerzeroscan (LayerZero being the underlying protocol of Stargate, the protocol used for this swap), it is still polled as pending by Li.Fi after 10 minutes (and probably up to infinity), see:

https://layerzeroscan.com/111/address/0x701a95707a0290ac8b90b3719e8ee5b210360883/message/106/address/0x9d1b1669c73b033dfe47ae5a0164ab96df25b944/nonce/92470
image

Again, refresh made things happy here.

File: src/components/Trade/hooks/useSwapper/useSwapper.tsx [hunk 4]
```diff
@@ -290,6 +290,7 @@ export const useSwapper = () => {
 
       if (!activeQuote) throw new Error('no activeQuote available')
       if (!wallet) throw new Error('no wallet available')
+      if (!supportsETH(wallet)) throw Error('eth wallet required')
```
Comment (gomesalexandre) [inline src/components/Trade/hooks/useSwapper/useSwapper.tsx:293]: tyvm

File: src/lib/swapper/swappers/LifiSwapper/utils/getNetworkFeeCryptoBaseUnit/getNetworkFeeCryptoBaseUnit.ts [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    throw new SwapError('[getNetworkFeeCryptoBaseUnit] invalid chain adapter', {
+      code: SwapErrorType.TRADE_QUOTE_FAILED,
+      details: { chainId },
+    })
+  }
+
+  const { transactionRequest } = await lifi.getStepTransaction(lifiStep)
+  const { value, from, to, data, gasLimit, gasPrice, maxFeePerGas } = transactionRequest ?? {}
+
+  if (value === undefined || from === undefined || to === undefined || data === undefined) {
```
Comment (gomesalexandre) [inline src/lib/swapper/swappers/LifiSwapper/utils/getNetworkFeeCryptoBaseUnit/getNetworkFeeCryptoBaseUnit.ts:33]: We're dealing with `string | number | bigint | BigNumber | Bytes | undefined` type here, I believe we may be able to use the `!` operator on most of these - may have to double check whether 0-value is returned as `0`, `'0'`, a 0 BN instance though

File: src/lib/swapper/swappers/LifiSwapper/utils/getNetworkFeeCryptoBaseUnit/getNetworkFeeCryptoBaseUnit.ts [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+  const gasFee = maxFeePerGas
+    ? bn(maxFeePerGas as string)
+    : bnOrZero(gasLimit as string).times(gasPrice as string)
+
+  // TEMP: this is necessary because infura currently cannot estimate gas for some lifi contract
+  // interactions, so instead of calling our existing stack which relies on infura we call the
+  // optimism gas oracle directly.
+  if (chainId === KnownChainIds.OptimismMainnet) {
+    const optimismGasOracleAddress = '0x420000000000000000000000000000000000000f'
```
Comment (gomesalexandre) [inline src/lib/swapper/swappers/LifiSwapper/utils/getNetworkFeeCryptoBaseUnit/getNetworkFeeCryptoBaseUnit.ts:47]: Nitpick, this is pretty well contained so not a big deal but do we want to abstract this as a `OPTIMISM_GAS_ORACLE_ADDRESS` constant outside this scope?

File: src/lib/swapper/swappers/LifiSwapper/utils/transformLifiFeeData/transformLifiFeeData.ts [hunk 7]
```diff
@@ -1,77 +1,57 @@
...
-        }
-      } else {
-        acc[assetId].amountCryptoBaseUnit = bn(acc[assetId].amountCryptoBaseUnit)
-          .plus(amount)
-          .toString()
+  const protocolFees = allPayableFeeCosts.reduce<Record<AssetId, ProtocolFee>>((acc, feeCost) => {
+    const { amount: amountCryptoBaseUnit, token } = feeCost
+    const assetId = lifiTokenToAssetId(token)
+    const asset = assets[assetId]
+    if (acc[assetId] === undefined) {
```
Comment (gomesalexandre) [inline src/lib/swapper/swappers/LifiSwapper/utils/transformLifiFeeData/transformLifiFeeData.ts:38]: `!` if you feel like it

**Review by gomesalexandre — APPROVED** [review]
Conceptually LGTM 🤌 - added a few non-blocking comments.

If the previously noted issue WRT max is related to my monkey patch (which it seems to be, according to your screenshots), feel free to merge

**Comment by woodenfurniture** [comment]
@gomesalexandre we're good to go, addressed your points where required.

> Paranoia "regression" (ish, since XDAI native asset mapping isn't currently present) test of a swap to OP via hop - no L1 fees since the fees are taken in XDAI on the sell side only

Confirmed this is resolved with https://github.com/shapeshift/web/pull/4616:
image

> If the previously noted issue WRT max is related to my monkey patch (which it seems to be, according to your screenshots), feel free to merge

Man, thank you for raising this one. There was a race condition in `TradeInput` that was using the wrong quote to calculate the max trade. Fix in 62c4e7d4945241ebc1a9b4ace164cda3edde1f62

> Another thing I've noticed and we may want to report to Li.Fi is despite the Tx being completed on Layerzeroscan (LayerZero being the underlying protocol of Stargate, the protocol used for this swap), it is still polled as pending by Li.Fi after 10 minutes (and probably up to infinity)

This one is tricky to solve without lifi changing implementation on their side. I'll likely address this in multi-hop with new state management which "races" lifi and our node - if tx is confirmed on our node stop polling lifi, otherwise keep polling lifi

**Review by 0xApotheosis — APPROVED** [review]
Gomes' excellent comments have been addressed, and they new fee threshold is much more sane than the arbitrary 99% hack we were previously using, big gm.

**Pull request merged**